librz/analysis/arch/ppc/ppc_il_ops.c

```diff
@@ -0,0 +1,1597 @@
+		op0 = UNSIGNED(128, op0);
+		op1 = UNSIGNED(128, op1);
+	} else if (ppc_is_mul_div_u(id) && !ppc_is_mul_div_d(id, mode)) {
+		op0 = UNSIGNED(64, op0);
+		op1 = UNSIGNED(64, op1);
+	} else if (!ppc_is_mul_div_u(id) && ppc_is_mul_div_d(id, mode)) {
+		op0 = SIGNED(128, op0);
+		op1 = SIGNED(128, op1);
+	} else {
+		op0 = SIGNED(64, op0);
+		op1 = SIGNED(64, op1);
+	}
+
+	switch (id) {
+	default:
+		NOT_IMPLEMENTED;
+	case PPC_INS_MULHD:
+	case PPC_INS_MULHDU:
+	case PPC_INS_MULHW:
+	case PPC_INS_MULHWU:
+	case PPC_INS_MULLI:
+	case PPC_INS_MULLD:
+	case PPC_INS_MULLW:
+		prod = MUL(op0, op1);
+		break;
+	case PPC_INS_DIVWU:
+	case PPC_INS_DIVDU:
+		prod = DIV(op0, op1);
+		break;
+	case PPC_INS_DIVW:
+	case PPC_INS_DIVD:
+		prod = SDIV(op0, op1);
+		break;
+	}
+
+	// Get high word/double word of product.
+	if (id == PPC_INS_MULHW || id == PPC_INS_MULHWU) {
+		prod = SHIFTR0(prod, U8(32));
+	} else if (id == PPC_INS_MULHD || id == PPC_INS_MULHDU) {
+		prod = SHIFTR0(prod, U8(64));
+	}
+
+	RzILOpEffect *cr0 = set_cr0 ? ppc_cmp_set_cr(VARG(rT), UA(0), true, "cr0", mode) : EMPTY();
+	return SEQ2(SETG(rT, CAST(PPC_ARCH_BITS, IL_FALSE, prod)), cr0);
+}
+
+static RzILOpEffect *move_from_to_spr_op(RZ_BORROW csh handle, RZ_BORROW cs_insn *insn, const cs_mode mode) {
+	rz_return_val_if_fail(handle && insn, EMPTY());
+	ut32 id = insn->id;
+
+	const char *rS = cs_reg_name(handle, INSOP(0).reg);
+	const char *rT = cs_reg_name(handle, INSOP(0).reg);
+	const char *spr_name;
+	// Some registers need to assemble the value before it is read or written (e.g. XER with all its bits).
+	// Leave it NULL if the value of the SPR or RS should be used.
+	RzILOpEffect *set_val = NULL;
+	// Size of the value written to the target register (SPR or RT)
+	ut32 size = PPC_ARCH_BITS;
+
+	switch (id) {
+	default:
+		NOT_IMPLEMENTED;
+	// ???
+	case PPC_INS_MTDCR:
+	case PPC_INS_MFDCR:
+		NOT_IMPLEMENTED;
+	case PPC_INS_MFVSCR:
+	case PPC_INS_MTVSCR:
+		NOT_IMPLEMENTED;
+		break;
+	case PPC_INS_MFMSR:
+		spr_name = "msr";
+		break;
+	case PPC_INS_MTMSR:
+	case PPC_INS_MTMSRD:
+		NOT_IMPLEMENTED;
+	case PPC_INS_MTCR:
+	case PPC_INS_MTCRF:;
+		ut32 mask = 0xffffffff;
+		if (id == PPC_INS_MTCRF) {
+			ut8 fxm = INSOP(0).imm;
+			rS = cs_reg_name(handle, INSOP(1).reg);
+			mask = ppc_fmx_to_mask(fxm);
+		}
+		RzILOpEffect *set_cr = SETL("cr", LOGAND(UNSIGNED(32, VARG(rS)), U32(mask)));
+		return SEQ2(set_cr, ppc_sync_crx_cr(false, mask));
+
+	case PPC_INS_MFCR:
+		return SEQ2(ppc_sync_crx_cr(true, 0x0), SETG(rT, EXTZ(VARL("cr"))));
+
+	// Note: We do not update CR after the OCRF operations.
+	case PPC_INS_MTOCRF:
+	case PPC_INS_MFOCRF:;
+		// This instruction is already implemented but was broken at the time.
+		// For the implementation see: https://github.com/Rot127/rizin/tree/ppc-rzil-broken-insn-impl
+		//
+		// Bug:
+		// Capstone v5 does not store the fxm value in the operands.
+		// See: https://github.com/capstone-engine/capstone/issues/1903
```

Comment:
Could you please list all capstone bugs you met in the PR description?

Reply:
Updated the description.